### PR TITLE
feat(eslint-plugin)!: add rule `consistent-type-assertions`

### LINKED ISSUE #332: Rule suggestion: generalize @typescript-eslint/no-angle-bracket-type-assertion
I would like to have a more general version of `@typescript-eslint/no-angle-bracket-type-assertion`, maybe called `type-assertion`. It would have two options, that can be enabled independently:

`no-angle-bracket`: If this is enabled, angle bracket type assertions are disallowed.
`no-as`: If this is enabled, `as` type assertions are disallowed.

This would allow different use cases. Some people like to use the old assertions instead of the new ones. Personally, I would like to lint all type casting since they introduce unsafety.

The default behaviour would be `['no-angle-bracket']` to match the current situation. If no options are enabled, the lint could warn that the lint can be disabled.

### LINKED ISSUE #668: [no-object-literal-type-assertion] make allowAsParameter work with "throw" keyword?
```JSON
{
  "rules": {
    "@typescript-eslint/no-object-literal-type-assertion": ["error", {
      "allowAsParameter": true 
    }]
  }
}
```
My code:
```TS
else throw { code: 'E42', message: 'foo failed' } as CustomError;
```

**Expected Result**
I hope eslint would allow it, since I've set `"allowAsParameter": true`

**Actual Result**
It's not allowed, because `throw` is not a function, it's a function-like keyword, just like `typeof` or `delete`.

Could we make it treat `throw` as a function in regards to `"allowAsParameter": true` ?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.9.0` |

### PR DESCRIPTION
**Pull request by bradzacher**
BREAKING CHANGE: Merges both `no-angle-bracket-type-assertion` and `no-object-literal-type-assertion` into one rule

Fixes #332
Fixes #668

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/consistent-type-assertions.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Enforces consistent usage of type assertions. (consistent-type-assertions)
+
+## Rule Details
+
+This rule aims to standardise the use of type assertion style across the codebase.
+
```
Comment (JamesHenry) [inline packages/eslint-plugin/docs/rules/consistent-type-assertions.md]: You start out consistently using "type assertions" (the more official term), but then use "cast" (the commonly used, but less strictly accurate term) a few times as well.

Please could you add a short explanation in the intro about this fact/relationship?

e.g. something like:

> Type assertions are also commonly referred as "type casting" in TypeScript (even though it is technically slightly different to what is understood by type casting in other languages), so you can think of type assertions and type casting referring to the same thing. It is essentially you saying to the TypeScript compiler, "in this case, I know better than you!"

File: packages/eslint-plugin/src/rules/no-angle-bracket-type-assertion.ts [hunk 2]
```diff
@@ -15,6 +15,8 @@ export default util.createRule({
         "Prefer 'as {{cast}}' instead of '<{{cast}}>' when doing type assertions.",
     },
     schema: [],
+    deprecated: true,
```
Comment (JamesHenry) [inline packages/eslint-plugin/src/rules/no-angle-bracket-type-assertion.ts]: As we applying this as a breaking change should we just remove the rules? How long were you planning on keeping the deprecated implementations for?

File: packages/eslint-plugin/src/rules/no-object-literal-type-assertion.ts [hunk 3]
```diff
@@ -36,6 +36,8 @@ export default util.createRule<Options, MessageIds>({
         },
       },
     ],
+    deprecated: true,
```
Comment (JamesHenry) [inline packages/eslint-plugin/src/rules/no-object-literal-type-assertion.ts]: Same as above

File: packages/eslint-plugin/src/rules/no-angle-bracket-type-assertion.ts (hunk 2 above)
Comment (JamesHenry) [reply packages/eslint-plugin/src/rules/no-angle-bracket-type-assertion.ts]: Just saw #732 which answers that question!

Let's just get it all done in this PR now that we are into v2 prereleases

File: packages/eslint-plugin/src/rules/no-object-literal-type-assertion.ts (hunk 3 above)
Comment (JamesHenry) [reply packages/eslint-plugin/src/rules/no-object-literal-type-assertion.ts]: Answered

**Comment by JamesHenry** [comment]
@bradzacher Linting failed on this one, looks like we are still using the deleted rules as of the current commit, so it is failing to find them

**Comment by bradzacher** [comment]
oops forgot about the recommended config

**Pull request merged**